libguard/guard_exception.hpp

```diff
@@ -0,0 +1,100 @@
+#pragma once
+
+#include <exception>
+#include <string>
+
+class guardFileOpen : public std::exception
```

Comment:
Why not have a common base class for all your guard exceptions? 
So, that catcher no need to worry about write multiple catches if he going to take some action?

still, you can argue why not caller use std::exception? this is a standard exception and the catcher maybe do a special action for all guard exceptions and different action for standard exception

Reply:
In that case how will user differentiate between the failures ?

Reply:
> Why not have a common base class for all your guard exceptions?
> So, that catcher no need to worry about write multiple catches if he going to take some common action?
> 
> still, you can argue why not caller use std::exception? this is a standard exception and the catcher maybe do a special action for all guard exceptions and different activities for standard exception

+1

Reply:
> In that case how will user differentiate between the failures ?

you can define custon structre , which includes the required error data.

struct error_info                                                               
{                                                                               
   ERR_TYPE err_type;                                                           
   const char *msg;                                                             
   void const*  private_data;                         
};

Reply:
Not really this will lead to more changes as compared to adding exceptions, as we need to change the return type of every function also changes in tool side. Also we thought of opting for CPP way and writing custom exceptions.